source/agora/consensus/Validation.d

```diff
@@ -990,6 +993,68 @@ public string isInvalidReason (const ref Block block, in ulong prev_height,
     return null;
 }
 
+/*******************************************************************************
+
+    Check the validator of an enrollment.
+
+    A Validator's enrollment is considered valid if:
+        - UTXO is unspent frozen utxo
+        - Signatures are authentic
+        - The frozen amount must be equal to or greater than 40,000 BOA
+        - The block height must be at least unlock_height or greater of the UTXO.
+
+    Params:
+        block_height = the height of the direct ancestor of this block
+        enrollment = The enrollment of the target to be verified
+        findUTXO = delegate to find the referenced unspent UTXOs with
+
+    Returns:
+        `null` if the validator's UTXO is valid, otherwise a string
+        explaining the reason it is invalid.
+
+*******************************************************************************/
+
+public string isInvalidEnrollmentReason (const ulong block_height,
+    const ref Enrollment enrollment, UTXOFinder findUTXO) nothrow @safe
+{
+    UTXOSetValue utxo_set_value;
+    if (!findUTXO(enrollment.utxo_key, size_t.max, utxo_set_value))
+        return "Unspent frozen UTXO not found for the validator.";
+
+    if (utxo_set_value.type != TxType.Freeze)
+        return "UTXO is not frozen.";
+
+    Point address;
+    try
+    {
+        address = Point(utxo_set_value.output.address);
+    }
+    catch (Exception ex)
+    {
+        return "Error converting address to point";
+    }
+
+    if (!() @trusted { return verify(address, enrollment.enroll_sig,
+        enrollment); }())
+        return "Enrollment signature verification has an error.";
+
+    if (utxo_set_value.output.value.integral() < Amount.MinFreezeAmount.integral())
+        return "The frozen amount must be equal to or greater than 40,000 BOA.";
```

Comment:
How about using the code `Amount.MinFreezeAmount.integral().to!string` instead of `40,000` as follows because the MinFreezeAmount might be changed.
```
return "The frozen amount must be equal to or greater than " ~
        Amount.MinFreezeAmount.integral().to!string ~ " BOA.";
```

Reply:
@Geod24 do you know if the compiler will try to CTFE the above? Or will this allocate a new string every time? If it allocates, we could work around it by assigning the string to a static variable though.

Reply:
We *could* do that, but not exactly this way. 
As @AndrejMitrovic hinted, this will most likely allocate.
To force CTFE, do:
```D
if (....)
{
    static immutable Message = "The frozen amount" ...;
    return Message;
}
```

It is paramount that error path do not allocate memory (or that this memory be freed). The reason is simple: If sending an invalid transaction to a node leads to memory allocation, one could easily sends many of those to increase the workload on the validator (as it would stress the garbage collector), which is a form of denial of service attack.

Reply:
CTFE: https://tour.dlang.org/tour/kr/gems/compile-time-function-evaluation-ctfe